src/cast_core/src/RandomSampler.cpp

```diff
@@ -0,0 +1,99 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/RandomSampler.hpp>
+#include <cast_core/actors/CollectionScanner.hpp>
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+
+namespace genny::actor {
+
+struct RandomSampler::PhaseConfig {
+    mongocxx::database database;
+    std::vector<std::string> collectionNames;
+    genny::DefaultRandom random;
+    std::uniform_int_distribution<> integerDistribution;
+    metrics::Operation readOperation;
+    metrics::Operation readWithScanOperation;
+    PhaseConfig(PhaseContext& context, const mongocxx::database& db, ActorId id, int collectionCount,
+        int threads, ActorCounter &counter)
+        : database{db}, readOperation{context.operation("Read", id)},
+        readWithScanOperation{context.operation("ReadWithScan", id)} {
+        int actorId = counter;
+        counter ++;
+        // Distribute the collections among the actors.
+        collectionNames = CollectionScanner::getCollectionNames(collectionCount, threads, actorId);
+        // Setup the int distribution.
+        integerDistribution = std::uniform_int_distribution(0, (int)collectionNames.size() - 1);
+    }
+};
+
+void RandomSampler::run() {
+    for (auto&& config : _loop) {
+        // Construct basic pipeline for retrieving one random record.
+        mongocxx::pipeline pipeline{};
+        pipeline.sample(10);
+        for (const auto&& _ : config) {
+            try {
+                auto statTracker = _collectionScannerCounter > 0 ? config->readWithScanOperation.start() :
+                  config->readOperation.start();
+                int index = config->collectionNames.size() > 1 ? config->integerDistribution(config->random) : 0;
+                auto cursor = config->database[config->collectionNames[index]].aggregate(pipeline,
+                    mongocxx::options::aggregate{});
+                for (auto doc : cursor){
+                  ;
+                }
+                statTracker.success();
+            } catch(mongocxx::operation_exception& e) {
+                BOOST_LOG_TRIVIAL(error) << boost::diagnostic_information(e);
```

Comment:
I don't know enough about the driver to know when this exception would be thrown but it seems like it may be just a normal course of business (the exception is expected) in which case I don't know if logging it is all that helpful.

I'd either

1. log it at debug level (which is off by default iirc) or
2. re-throw it (after calling `tracker.failure()`) if you're not expecting this exception to happen in the normal flow of the workload. Rethrowing it will cause the workload to end asap and will report the workload as a system failure in evergreen. (Syntax to re-throw is just `throw;`..I learned that only somewhat recently lol)

And you can catch by const-ref `catch(const mongocxx::operation_exception& e)`.

I think the same likely applies to the other Actor as well.